src/create/schema/index.ts

```diff
@@ -211,14 +212,45 @@ export const createSchema = <
   return createManualTypeSchema(zodSchema);
 };
 
+export const createSchema = <
+  Output = any,
+  Def extends ZodTypeDef = ZodTypeDef,
+  Input = Output,
+>(
+  zodSchema: ZodType<Output, Def, Input>,
+  state: SchemaState,
+): oas31.SchemaObject | oas31.ReferenceObject => {
+  if (state.visited?.has(zodSchema)) {
+    throw new Error(
+      `The schema at ${
+        state.path?.join(' > ') || '<root>'
+      } needs to be registered because it's circularly referenced`,
+    );
+  }
+  state.visited ??= new Set();
+  state.visited.add(zodSchema);
+  const result = createSchemaSwitch(zodSchema, state);
+  state.visited.delete(zodSchema);
+  return result;
+};
+
 export const createSchemaOrRef = <
   Output = any,
   Def extends ZodTypeDef = ZodTypeDef,
   Input = Output,
 >(
   zodSchema: ZodType<Output, Def, Input>,
   state: SchemaState,
+  subpath?: string[],
 ): oas31.ReferenceObject | oas31.SchemaObject => {
+  if (subpath) {
+    const path = (state.path ??= []);
+    const lengthBefore = path.length;
+    path.push(...subpath);
+    const result = createSchemaOrRef(zodSchema, state);
+    path.length = lengthBefore;
```

Comment:
Hmmm I wonder if instead of this would it be better if we initialised a new path array every time we traverse schemas?

Reply:
Well, there's even Immutable.js for more performant immutable sets and lists and such :D But I dunno, this is the way I would keep it for now but I don't have a strong opinion